cpp/src/arrow/compare.cc

```diff
@@ -364,10 +426,15 @@ static bool IsEqualPrimitive(const PrimitiveArray& left, const PrimitiveArray& r
   }
 }
 
+// A bit confusing: ArrayEqualsVisitor inherits from RangeEqualsVisitor but
+// doesn't share the same preconditions.
+// When RangeEqualsVisitor is called, we only know the range sizes equal.
+// When ArrayEqualsVisitor is called, we know the sizes and null bitmaps are equal.
```

Comment:
It's a good point. I don't know that this was necessarily deliberate, but if you are interested in seeing it fixed we can do so

Reply:
One simple possibility is for `RangeEquals` to be implemented by slicing the inputs... It might be a tad more expensive for small inputs (allocating new vectors and shared pointers for the buffers).